jobrunner/executors/volumes.py

```diff
@@ -1,37 +1,127 @@
+    with atomic_writer(dest) as tmp:
+        shutil.copy(source, tmp, follow_symlinks=follow_symlinks)
+
+
+def docker_volume_name(job):
     return f"os-volume-{job.id}"
 
 
 class DockerVolumeAPI:
     def volume_name(job):
-        return volume_name(job)
+        return docker_volume_name(job)
 
     def create_volume(job, labels=None):
-        docker.create_volume(volume_name(job))
+        docker.create_volume(docker_volume_name(job))
 
     def volume_exists(job):
-        return docker.volume_exists(volume_name(job))
+        return docker.volume_exists(docker_volume_name(job))
 
     def copy_to_volume(job, src, dst, timeout=None):
-        docker.copy_to_volume(volume_name(job), src, dst, timeout)
+        docker.copy_to_volume(docker_volume_name(job), src, dst, timeout)
 
     def copy_from_volume(job, src, dst, timeout=None):
-        docker.copy_from_volume(volume_name(job), src, dst, timeout)
+        docker.copy_from_volume(docker_volume_name(job), src, dst, timeout)
 
     def delete_volume(job):
-        docker.delete_volume(volume_name(job))
+        docker.delete_volume(docker_volume_name(job))
 
     def touch_file(job, path, timeout=None):
-        docker.touch_file(volume_name(job), path, timeout)
+        docker.touch_file(docker_volume_name(job), path, timeout)
 
     def glob_volume_files(job):
-        return docker.glob_volume_files(volume_name(job), job.output_spec.keys())
+        return docker.glob_volume_files(docker_volume_name(job), job.output_spec.keys())
 
     def find_newer_files(job, path):
-        return docker.find_newer_files(volume_name(job), path)
+        return docker.find_newer_files(docker_volume_name(job), path)
+
+
+def host_volume_path(job):
+    path = config.HIGH_PRIVACY_VOLUME_DIR / job.id
+    path.parent.mkdir(parents=True, exist_ok=True)
+    return path
+
+
+class BindMountVolumeAPI:
+    def volume_name(job):
+        return host_volume_path(job)
+
+    def create_volume(job, labels=None):
+        host_volume_path(job).mkdir()
+
+    def volume_exists(job):
+        return host_volume_path(job).exists()
+
+    def copy_to_volume(job, src, dst, timeout=None):
+        # We don't respect the timeout.
+        volume = host_volume_path(job)
+        if src.is_dir():
+            shutil.copytree(
+                src,
+                volume / dst,
+                symlinks=True,
+                copy_function=copy_file,
+                dirs_exist_ok=True,
+            )
+        else:
+            copy_file(src, volume / dst)
+
+    def copy_from_volume(job, src, dst, timeout=None):
+        # this is only use to copy final outputs/logs. So we optimise by doing
+        # a rename. If the volume directory and the workspace directory are on
+        # the same filesystem, this should mean zero copying.
+        # Note: We don't respect the timeout.
+        #
+        path = host_volume_path(job) / src
+        dst.parent.mkdir(parents=True, exist_ok=True)
+        path.rename(dst)
```

Comment:
This isn't idempotent, and it means that the whole finalize step is no longer idempotent which I don't think we can tolerate. I think it would be possible to make it idempotent via some sort of tombstone file trickery but the details get a bit fiddly. It might be easiest in the first instance to make this a copy, which will still be a win as it's local, and then worry about how to do moves idempotently later.

Reply:
I have updated with a straight copy.

We of course relying on the duplication aspects of copying to provide various system properties like this. But we way the cost in two ways: latency and capacity. This move to faster coping should help with the latency, but not with the capacity cost. That's the big win that CoW and/or `rename` approach gains us.